### ISSUE TITLE
Unable to build multi-module project with fmp 3.5.34

### ISSUE THREAD

**Issue by geoand** [fabric8io/fabric8-maven-plugin#1184]
### Description

When trying to build [this](https://github.com/snowdrop/spring-boot-configmap-booster/tree/sb-1.5.10) multi module project with fmp `3.5.34` I get the following error

```
[ERROR] Failed to execute goal io.fabric8:fabric8-maven-plugin:3.5.34:build (fmp) on project spring-boot-configmap-parent: Failed to execute the build: Unable to create the build archive: Failed to create assembly for docker image  (with mode 'dir'): Error creating assembly archive docker: You must set at least one file.. If you include the build artifact please ensure that you have built the artifact before with 'mvn package' (should be available in the target/ dir). Please see the documentation (section "Assembly") for more information.
```

I should not that when building other non multi-module projects, everything worked just fine.

In order to bypass the problem, I downgraded to fmp `3.5.33`, where everything worked as expected

### Info

* f-m-p version : 3.5.34
* Maven version (`mvn -v`) : 3.5.2
```

```
* Kubernetes / OpenShift setup and version : Openshift 3.7
* If it's a bug, how to reproduce : Run mvn clean verify -Popenshift,openshift-it on the sb-1.5.10 brach

### Notes
You can easily switch versions of fmp in the project by setting the `fabric8-maven-plugin.version` Maven property

**Comment by piyush1594** [comment]
Hi @geoand 

I am not able to produce the bug. I may be doing something wrong. Can you please check or explain more clearly.

Here is what I did:

In file spring-boot-configmap-booster/tests/pom.xml

After the line 87 and 102, I added this

`<version>3.5.34</version>`

It looks like this now

```
<groupId>io.fabric8</groupId>
<artifactId>fabric8-maven-plugin</artifactId>
<version>3.5.34</version>
```
I am running openshift v3.7.0

and after that from root(spring-boot-config-map) folder, I hit

`mvn clean verify -Popenshift,openshift-it`

Then it works fine for me like project is build, it says BUILD SECCESS.

```
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-failsafe-plugin:2.20:verify (default) @ spring-boot-configmap-tests ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Spring Boot - ConfigMap Booster .................... SUCCESS [  5.970 s]
[INFO] Spring Boot - ConfigMap Booster - Greeting Service . SUCCESS [ 36.750 s]
[INFO] Spring Boot - ConfigMap Booster - Tests ............ SUCCESS [ 50.904 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:33 min
[INFO] Finished at: 2018-02-14T20:55:38+05:30
[INFO] Final Memory: 72M/668M
[INFO] ------------------------------------------------------------------------
```

Can you please check and reply back if i am doing something wrong or explain more clearly your issue.

**Comment by geoand** [comment]
Hello @piyush1594,

The easiest way to reproduce the issue is to make this single modification to the project:

Add the following lines in `pom.xml` (the pom in the root directory, not the child ones)

```xml
<properties>
  <fabric8-maven-plugin.version>3.5.34</fabric8-maven-plugin.version>  
</properties>
```

Please let me know if I can help any further.

Regards

**Comment by geoand** [comment]
@piyush1594 You can also reproduce the problem by not making any modifications at all to the project and issuing:

`mvn clean verify  -Dfabric8-maven-plugin.version=3.5.34 -Popenshift,openshift-it` from the root

**Comment by piyush1594** [comment]
Hey @geoand 

I tried building this project both using FMP 3.5.34 and FMP 3.5.33

Project build 3 module when I hit the command `mvn clean verify -Popenshift,openshift-it`

So for FMP 3.5.34 first is getting failed for Opesnhift version 3.6.0, 3.7.0 and 3.7.1

And for FMP 3.5.33 second is getting failed for Openshift version 3.6.0, 3.7.0 and 3.7.1

So this problem is not specific to openshift. For every openshift version, it is giving the same error for a single specified version.

Thanks

**Comment by geoand** [comment]
Hello @piyush1594,

That is good information to know. I didn't test with fmp `3.5.33`, but I know the project was working when we were using `3.5.30`.
I also checked with `3.5.35` and the issue still exists.

Thanks

**Comment by metacosm** [comment]
Hi @piyush1594,

Anyway to prioritize this issue? This is really blocking for RHOAR as we need OpenShift 3.7+ support *and* multi-module support ASAP… Do you have an ETA on availability of a fix for this (and/or new FMP release)?

Thanks in advance.

**Comment by rohanKanojia** [comment]
I bisected this issue; This seems to have begun at the [e69dfc9](https://github.com/fabric8io/fabric8-maven-plugin/commit/e69dfc902b22a8119849be24393c7001ed3a57c9)
Adding cc @rhuss, Could you possibly take a look at this one and share your thoughts?
Thanks.
 
Bisect on linux.
```
e69dfc902b22a8119849be24393c7001ed3a57c9 is the first bad commit
commit e69dfc902b22a8119849be24393c7001ed3a57c9
Author: Roland Huß <roland@ro14nd.de>
Date:   Fri Nov 24 10:54:38 2017 +0100

    feature: Don't use a default for skipBuildPom
    
    Instead, try to detect the default by checking whether there has been
    some image configurations with build sections resolved.
    
    This can always overwritten by explicitly specifying fabric8.skip.build.pom

:040000 040000 d02414af604d22fe851f8addb2bc5520f8a21efd 5a2e37051721387c1484fd3d86336856ce55778b M	plugin
```

Git bisect log
```
git bisect start
# good: [fad3a38f521005b25dddc5dd4236348bc9ace9b0] Fixes #1040 by turning validation errors into warnings
git bisect good fad3a38f521005b25dddc5dd4236348bc9ace9b0
# bad: [533f8445eb143425928ad7a2f469cb2bab4438ec] Merge pull request #1101 from apupier/1100-FixStartupScriptForWindows
git bisect bad 533f8445eb143425928ad7a2f469cb2bab4438ec
# good: [bc48668dff0bdb91103a92ceb0ffb9a8d47a0e27] Fix for stall build; let's remove PodWatcher and try with normal sleep()
git bisect good bc48668dff0bdb91103a92ceb0ffb9a8d47a0e27
# bad: [89abcae3019d2bcbd05142f83353165e581a49b2] More Refactoring of Test with proper naming and Exception Handling
git bisect bad 89abcae3019d2bcbd05142f83353165e581a49b2
# good: [7b82d5d7d0d022ef6d699ba205df2582e58dfb8b] Merge pull request #1115 from rohanKanojia/ISS-1111/os-regression-test
git bisect good 7b82d5d7d0d022ef6d699ba205df2582e58dfb8b
# bad: [fd9bdfe01a87a86f7fa331180b85e5ee5809992b] Added Test for Container Handler
git bisect bad fd9bdfe01a87a86f7fa331180b85e5ee5809992b
# bad: [f6fb470aea627dfc1a182c4c71f92489fe1ca6bb] Merge pull request #1112 from rhuss/pr/skip-pom-refined
git bisect bad f6fb470aea627dfc1a182c4c71f92489fe1ca6bb
# bad: [a5a72194e8789812a048145e1fffc46839950c56] fix: Remove bogus import and update documentation for skipBuildPom
git bisect bad a5a72194e8789812a048145e1fffc46839950c56
# bad: [e69dfc902b22a8119849be24393c7001ed3a57c9] feature: Don't use a default for skipBuildPom
git bisect bad e69dfc902b22a8119849be24393c7001ed3a57c9
# first bad commit: [e69dfc902b22a8119849be24393c7001ed3a57c9] feature: Don't use a default for skipBuildPom
```

**Comment by piyush1594** [comment]
Hi @geoand 

I just locally checked by building the above-mentioned project using FMP 3.5.34 You need to just change the command. You need to add one more flag for this.

`mvn clean -Dfabric8.skip.build.pom=true verify -Popenshift,openshift-it`

Before FMP 3.5.34 this flag was defaulted to true, so it was working fine. Now default value for this flag is not set. You need to set this flag to true while executing the command.

So for FMP 3.5.34, above command will work I just checked locally and for FMP 3.5.33, both the commands will work as it was default set to true.

Hope this helps. Let us know if this does not resolve your problem. Please give this a try.

Thanks

**Comment by geoand** [comment]
Hello @piyush1594,

Although the addition of `-Dfabric8.skip.build.pom=true` does allow the project to build, it doesn't seem to reasonable to expect users of FMP to have to set this flag for all their multi-module projects.
For one thing, existing CI setups (that worked up until `3.5.33`) now fail just by upgrading to FMP 3.5.34.

Regards

**Comment by piyush1594** [comment]
Hey @geoand,

After your comment, I just checked locally once again, the project is successfully getting build for me using the above command. I don't know why it's not working for you. I am using OpenShift 3.7.1 and FMP 3.5.34

Can you please share the updated logs?

Thanks

**Comment by metacosm** [comment]
A breaking change in default behavior in a minor release is definitely not expected. Breaking changes should be reserved to major versions.
@geoand we could mitigate the issue by adding the flag to our parent POM… if that's really the issue.

**Comment by geoand** [comment]
Hello again,

Perhaps I didn't phrase my response correctly, I will make another attempt :).

When I execute `mvn clean -Dfabric8.skip.build.pom=true verify -Popenshift,openshift-it`, everything works correctly as you mention.

What I was is saying, is that it doesn't make much such to force users of FMP to set a (seemingly) unrelated setting in order to get their multi-module projects to function correctly.
I don't have any knowledge of the internals of FMP, but it seems to me much more straightforward and intuitive to just have `mvn clean verify -Popenshift,openshift-it` working for both multi-module and non-multi-module projects. 
A couple of examples that reinforce my aforementioned argument are the following:
* If I am new to FMP and I see that my multi-module project is failing, there is very little chance that I will discover that I need to set `fabric8.skip.build.pom` to true. In that case I might try using older versions of FMP, or perhaps give up on FMP entirely...
* If I have an existing multi-module project that used FMP version 3.5.33 and everything was working just fine (locally and in my CI setup), then simply upgrading to FMP 3.5.34 or 3.5.35 will break my project for reasons that are completely opaque to me... The most likely solution in that case would be for me to simply rollback to using 3.5.33 and not go through the trouble of looking for the "correct" fix.
Furthermore in this case, even if as a developer I find that I need to set `fabric8.skip.build.pom` to true, I might not be in control of the CI environment the project is built on, rendering me incapable of using new versions of FMP.

Thank you for your time and I hope that my arguments have become clearer :)

Regards

**Comment by geoand** [comment]
@metacosm I have tested the use of the flag and it does indeed fix the issue.

However even if I do add the flag, our CI will still not work correctly due to #1183 (which won't be the case once we upgrade to Openshift 3.7).

So for the time being @metacosm I propose that we don't make any changes to our POM, since I believe it will only increase our confusion about the state of our CI. Once we upgrade to Openshift 3.7, then we can safely upgrade to FMP 3.5.35 and set to necessary flag to get multi-module projects working

**Comment by piyush1594** [comment]
Hi @geoand 

This is mentioned here https://maven.fabric8.io/#build-goal-configuration in Table:10 Global Configuration. 

I understand and totally agree with your concern. I really appreciate your feedback. We have already started maintaining CHANGELOG.MD. We should release a note along fmp for changes. We will try our best to implement those things soon like Matrix Table, Release Notes, CHANGELOG.MD,proper documentation update. 

Thanks for this.

**Comment by geoand** [comment]
Thank you very much for the update!

Regards

**Comment by piyush1594** [comment]
Hey @geoand 

Can we close this issue now?

**Comment by geoand** [comment]
Hello @piyush1594,

That's up to you :). 
I personally think that the default behavior should be reverted to what it was in FMP 3.5.33, but that's just my opinion.

**Comment by metacosm** [comment]
I agree: breaking changes should be reserved to major versions and shouldn’t occur on micro versions (as is the case here).

**Comment by metacosm** [comment]
Hi @piyush1594,

With more consideration of the case, there are 2 things that are problematic:
1. FMP should be working out of the box for multi-module projects without having to specify an option
2. The option is not clearly documented: you mention `fabric8.skip.build.pom` and a global configuration table but that property doesn't exist in that table. The closest one is `docker.skip.build.pom` so there currently is no way for a user to discover that property on their own. Moreover, the property is poorly documented: why such a property is needed at all?

So, from our perspective, while we definitely can work around this issue for the time being, the current solution (i.e. change your build and look at the documentation where the property is actually not documented) is not acceptable and deserves proper handling. So, no, this issue shouldn't be closed.

**Comment by Ladicek** [comment]
I strongly agree that this breaking change should be rolled back.

**Comment by cmoulliard** [comment]
This issue is blocking for RHOAR Boosters and also OSIO users developing multi-module projects.

**Comment by maxandersen** [comment]
> I strongly agree that this breaking change should be rolled back.

do we have a pointer to which commit/PR actually changed the behavior ?

**Comment by maxandersen** [comment]
@metacosm @cmoulliard i agree its a bad change, but can you explain to me why you cannot set the property in boosters ?  then it would at least work for new projects; but old ones would still be broken.

**Comment by piyush1594** [comment]
Hey @maxandersen 

This comes into effect after this https://github.com/fabric8io/fabric8-maven-plugin/pull/1112

If someone sets this flag, it will work for previous versions also. 

If someone does not specify the flag then it will work until 3.5.33

Thanks

**Comment by Ladicek** [comment]
@maxandersen FYI, I filed https://github.com/openshiftio/booster-parent/issues/22 to address this for all boosters in one place.

**Comment by metacosm** [comment]
@maxandersen we can workaround the issue indeed by setting the property. However, this is a major breaking change in a micro release, a change that's not clearly documented. Moreover, it doesn't "feel" right that one should pass a property to work with multi-module projects (which is far from being the exception) from now on.

**Comment by rhuss** [comment]
Let me just briefly explain the motivation for the original change which was not meant to break backwards compatibility.

For projects of type `pom` by default fabric8-maven-plugin is always skipped. However, there are very [valid uses cases](https://github.com/syndesisio/syndesis/blob/master/app/s2i/pom.xml) which define a fabric8-maven-plugin configuration for only building images (without compiling). In order to no skip such projects the default behaviour was changed, that the fmp run is skipped in `pom` typed projects only when *no  fmp build configuration is attached. The rationale is that, *if* you configure the plugin, then you *expect* it to run (or the other way round, people were suprised that it has been not run although a plugin has been configured). Having the plugin be skipped by default is very confusing and caused several hour of debugging.

So, it should be backwards compatible _if_ your pom does not contain a build configuration in the fmp plugin configuration. If used as default for multi module children, the configuration should go in a `<pluginManagement>` section anyway (and not in a `<plugin>` configuration directly), so that the build is skipped like before.

Actually, I'm happy to revert this behaviour, if we add it to a minor release change (e.g. for 3.6) as it is really counter intuitive, if you configure a f-m-p build in a `pom` typed project (e.g. because this project *only* for building, nothing else) and then the build doesn't work because 'skip' is the default. 

But maybe it would also work for you, too, to move the default fmp build configuration in a parent pom.xml from a `<plugin>` to a `<pluginManagement>` section ? (which I would do anyway).

Let me know, how you would like to proceed ...

**Comment by metacosm** [comment]
I understand there are valid use cases for that change but shouldn't these use cases be the one that need to change as opposed to breaking things that were working with previous minor releases? In any case, I'd be in favor of reverting that change and deferring it to a minor release (at the very least). That new behavior should be also documented (probably by explaining as you just did).

Regarding the `pluginManagement` vs. `plugin` issue: it is doable. We already have a parent where the plugin is configured. We used `plugin` as opposed to `pluginManagement` because the plugin is used in exactly the same way in all children projects and so we wanted to avoid repeating the plugin declaration in all projects if there was no need to do so. Also, our parent is actually a child of another parent which *does* declare the plugin in a `pluginManagement` section.

Finally, personally, I find it weird to use Maven to declare a project as a pom just to build an external resource that has not much to do with a java artifact. Pom projects should be reserved to BOMs and parent definition, in my opinion, but I'm no Maven expert. So, supporting these use cases should be secondary to supporting projects which actually build java applications. But that's only my personal opinion :)

**Comment by maxandersen** [comment]
thanks @rhuss; +1 for changing behavior back in minor stream as it breaks existing projects.

@metacosm fully agree its a bad breaking change; just wanted to make sure you didn't block totally.

**Comment by metacosm** [comment]
@maxandersen we're probably going to go with the property-in-the-parent solution because this is the one that involves the less changes.

Just to re-iterate, though, this change seems (and, please, do correct me if I'm wrong) to be made to support fringe cases (i.e. no artifact projects) in Maven builds. So, to me, the solution to supporting such projects shouldn't force more traditional projects to change but rather provide a workaround for these projects (i.e. reverse the hierarchy of needing to work around an issue). Accordingly, my ideal solution would be to revert the change and figure out a way to support these other projects that doesn't impact more traditional projects instead of simply re-introducing that change in a minor release. I'd be happy to have my mind changed, though, because I might be missing the bigger picture.

**Comment by rhuss** [comment]
> I understand there are valid use cases for that change but shouldn't these use cases be the one that need to change as opposed to breaking things that were working with previous minor releases?

As mentioned, it was not meant to break backwards compatibility, and there has been a [full review](https://github.com/fabric8io/fabric8-maven-plugin/pull/1112) for this change. So please just consider this just as a bug, not as an intentional breakage of backwards compatibility (in contrast to the usual tradition of fmp in the past which never followed a semversion semantics).

> Just to re-iterate, though, this change seems (and, please, do correct me if I'm wrong) to be made to support fringe cases

Well, actually this property was introduced for a fringe use case, breaking a lot of builds at that time (without notice). Tbh, I consider it weird to declare a plugin, configure it and expect it to *not* run, regardless which packaging. That's what you are doing in your parent pom to save the switch-on of a plugin (which actually _I_ would find confusing, imo any plugin should be enabled explicitely in a submodule if used, as any used dependency should be declared directly and not relying on transitive dependencies. But your mileage may vary.

> @metacosm fully agree its a bad breaking change; just wanted to make sure you didn't block totally.

@maxandersen any idea how this can be prevented in the future to avoid bad guys doing bad things ? would be awesome.

I'm going to submit a reverting PR later today (or tomorrow morning).

**Comment by hrishin** [comment]
Adding a regression test for multimodule project
https://github.com/fabric8io/fabric8-maven-plugin/issues/1197

**Comment by piyush1594** [comment]
Hey @geoand 

We have reverted the change by #1213. This will be available in next release. And #1214 is there which will be included whenever we will do the minor release. Then this behavior will change and you have to add the flag.

Thanks for the issue.

**Comment by geoand** [comment]
Thank you @piyush1594 for the update!

**Comment by piyush1594** [comment]
Hey @geoand @metacosm 

We just released [FMP 3.5.38](https://repo.maven.apache.org/maven2/io/fabric8/fabric8-maven-plugin/3.5.38/). Please give it a try. Looking forward to your feedback.

Thanks

**Comment by geoand** [comment]
Hi @piyush1594 

Thank you for the update!
We'll try it and let you know.
